crates/ty_python_semantic/src/semantic_index/builder.rs

```diff
@@ -2654,3 +2654,76 @@ fn dunder_all_extend_argument(value: &ast::Expr) -> Option<&ast::Expr> {
 
     (attr == "__all__").then_some(value)
 }
+
+/// Builds an interval-map that matches expressions (by their node index) to their enclosing scopes.
+///
+/// The interval map is built in a two-step process because the expression ids are assigned in source order,
+/// but we visit the expressions in semantic order. Few expressions are registered out of order.
+///
+/// 1. build a point vector that maps node indices to their corresponding file scopes. The
+///    vector is sorted in ascending order of node indices and uses insertion sort to maintain the order.
+///    Insertion sort is efficient because most expressions are registered in order.
+/// 2. Condense the point vector into an interval map by collapsing adjacent node indices with the same scope
+///    into a single interval.
+struct ExpressionsScopeMapBuilder {
+    expression_and_scope: Vec<(NodeIndex, FileScopeId)>,
+}
+
+impl ExpressionsScopeMapBuilder {
+    fn new() -> Self {
+        Self {
+            expression_and_scope: vec![],
+        }
+    }
+
+    fn record_expression(&mut self, expression: &impl HasTrackedScope, scope: FileScopeId) {
+        let expression_index = expression.node_index().load();
+
+        if self
+            .expression_and_scope
+            .last()
+            .is_some_and(|last| last.0 <= expression_index)
+        {
+            self.expression_and_scope.push((expression_index, scope));
+            return;
+        }
+
+        let insertion_point = self
+            .expression_and_scope
+            .iter()
+            .rposition(|(index, _)| *index <= expression_index)
+            .map(|index| index + 1)
+            .unwrap_or_default();
+
+        self.expression_and_scope
+            .insert(insertion_point, (expression_index, scope));
+    }
+
+    fn build(self) -> ExpressionsScopeMap {
+        let mut iter = self.expression_and_scope.into_iter();
+        let Some(first) = iter.next() else {
+            return ExpressionsScopeMap::default();
+        };
+
+        let mut interval_map = Vec::new();
+
+        let mut current_scope = first.1;
+        let mut range = first.0..NodeIndex::from(first.0.as_u32() + 1);
+
+        for (index, scope) in iter {
+            if scope == current_scope {
+                range.end = NodeIndex::from(index.as_u32() + 1);
+                continue;
+            }
+
+            interval_map.push((range, current_scope));
+
+            current_scope = scope;
+            range = index..NodeIndex::from(index.as_u32() + 1);
```

Comment:
Not sure if better, but avoids wrapping/unwrapping of node indices into u32's (with theoretical overflow problems)
```suggestion
        let mut range = first.0..=first.0;

        for (index, scope) in iter {
            if scope == current_scope {
                range = (*range.start())..=index;
                continue;
            }

            interval_map.push((range, current_scope));

            current_scope = scope;
            range = index..=index;
```